nrf52840_pca10090 reset pin & HCI

lemrey opened the pull request:
Opened one PR for convenience.

1) add a GPIO reset pin to `nrf52840_pca10090`
2) update H4 driver to use the GPIO reset pin with `nrf52840_pca10090` controllers
3) update `samples/bluetooth/hci_uart` config for `nrf52840_pca10090` to enable the GPIO reset and NOP

@carlescufi 

Addresses: #15334.

lemrey committed:
boards: arm: nrf52840_pca10090: update log settings

Add a board-specific log level setting and change log levels.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
boards: arm: nrf52840_pca10090: add a GPIO reset line

Let one of the MCU interface pins be configured to act
as a reset line. This mitigates the lack of a connection
between the nRF9160 and nRF52840 PINRESET.
Minor refactoring and updated comments.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
Bluetooth: hci: h4: use GPIO reset for nrf52840_pca10090 controllers

This patch adds support to the nRF9160 for using a dedicated
GPIO pin to reset controllers running on nrf52840_pca10090.

It resets the controller before opening the H4 device, and it
delays the controller from booting until all bytes traveling
to the host have been received and drained from the UART,
thus ensuring that communication can begin from a clean state.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

lemrey committed:
samples: bluetooth: hci_uart: update nrf52840_pca10090 config

Enable the GPIO reset pin feature of nrf52840_pca10090
on pin P1.02 and the wait NOP feature of the controller.
These are used to let nrf9160_pca10090 synchronize the
HCI communication with nrf52840_pca10090 controllers.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

carlescufi commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -434,6 +437,10 @@ static int h4_open(void)
 	if (nrf51_init(h4_dev) < 0) {
 		return -EIO;
 	}
+#elif defined(CONFIG_BT_NRF52840_PCA10090_GPIO_RESET)
```
I would be OK with letting this one go through, but I really want @jhedberg to weigh in.
The alternative is to have a generic hook such as:

```
#if defined(CONFIG_BT_DRIVER_OPEN_HOOK)
    if (bt_driver_open_hook(h4_dev) < 0) {
        return -EIO;
    }
#endif
```

And then have the implementation of `bt_driver_open_hook()` in the `board.c` of the corresponding board.

carlescufi commented on drivers/bluetooth/Kconfig:
```diff
@@ -24,6 +24,27 @@ if BT_CUSTOM
+
+if BT_NRF52840_PCA10090_GPIO_RESET
+
+config BT_NRF52840_PCA10090_GPIO_RESET_PIN
```
This belongs in the board Kconfig, and in fact you already have it there, so please remove this option.

lemrey replied:
The host and the controller run on two different boards; the host does not have the macro selecting the pin defined (that's in `nrf52840_pca10090` board Kconfig), it needs its own.

carlescufi replied:
Right, but it still should be in the board Kconfig, not this one

galak commented:
Is this addressing #15334?

lemrey commented:
@galak yes, sorry I forgot to add it to the description, added now.

carlescufi committed:
drivers: Bluetooth: Generalize IC-specific setup hook

In order to generalize the currently specialized nRF51 IC setup hook,
make the following changes:

- Generalize the hook to bt_ic_setup()
- Use a weak NOP version by default
- Move the currently existing one to the board folder

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi reviewed (approved):
Tweaked it as described in the commit messages

anangl commented on boards/arm/nrf52840_pca10090/integrity.c:
```diff
@@ -49,3 +49,14 @@ BUILD_ASSERT_MSG(IS_ENABLED(CONFIG_BOARD_PCA10090_SWITCH0_PHY) ||
 		 IS_ENABLED(CONFIG_BOARD_PCA10090_SWITCH1_ARDUINO),
 		 "Invalid switch 2 routing");
+
+#if CONFIG_BOARD_PCA10090_NRF52840_RESET
```
This could be checked inside the `BUILD_ASSERT_MSG` as `!IS_ENABLED(CONFIG_BOARD_PCA10090_NRF52840_RESET)`. You won't need then to add this missing newline at the end of the file. ;-)

carlescufi replied:
I don't think that would work, I would need to negate all the other existing `IS_ENABLED()` which I think makes it more confusing. I will leave as-is but add the newline :)

carlescufi replied:
Fixed

anangl commented on boards/arm/nrf52840_pca10090/board.c:
```diff
@@ -53,12 +55,18 @@ LOG_MODULE_REGISTER(board_control_pca10090);
+/* The following tables specify the default configuration of each pin.
+ * The switches have active-low logic, so when a pin is set to 1 in
+ * these tables, pull it low to leave it set to its default.
+ * When it is set to 0, pull it high to activate the non-default setting.
```
I think this comment could be rephrased to be more comprehensible. Or is it just me that could not understand it despite reading it several times? Perhaps one of these "it" could be clarified, as well as the default stuff - it this "default configuration" and "non-default setting" referring to the same kind of thing?
Same applies to the very similar comment in the loop that configures the pins.

carlescufi replied:
Rewritten

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -423,20 +419,35 @@ static int h4_send(struct net_buf *buf)
+{
+	/* Return a value > 0 by default so that h4_open knows to discard
+	 * any bytes, since this function does nothing. */
+	return 32;
```
This return value meaning seems very specific to UART. I'm not sure it's possible to justify that this is an obvious choice regardless of the transport used (or that no other info needs to be returned by other transports). I wouldn't be that much bothered by this, but the way you've named the callback is completely generic and should apply to any transport. Would it be possible to make the return value void and then handle the discarding internally in the setup function?

jhedberg replied:
To put this another way: it should be possible to add calls to `bt_hci_transport_setup()` to all existing HCI drivers' `open()` callbacks in a meaningful way. In fact, it'd be nice if this PR did that for completeness.

carlescufi replied:
Yep, I can do that!

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -423,20 +419,35 @@ static int h4_send(struct net_buf *buf)
+  *
+  * @return number of additional bytes to discard, negative error
+  * 	    value on failure
+  */
```
Wouldn't it be good to put a prototype for `bt_hci_transport_setup()` and this documentation in a driver-agnostic header file? I suppose `include/drivers/bluetooth/hci_driver.h` would be the obvious choice.

carlescufi replied:
Agreed, will fix now

carlescufi commented:
@jhedberg comments addressed, thanks for the review and innput!
